### PR TITLE
Make aggregating annotation processors faster

### PR DESCRIPTION
**Pull request by oehme**
Aggregating processors need to reprocess all classes they are
interested in on any change. Until now this was ensured by passing
all their source files to the compiler for recompilation.

However, the Java compiler can run annotation processing on class
files too. So instead of recompiling the (unchanged) source files
we now tell the compiler which classes to reprocess. This adds a
small new limitation for aggregating processors - their annotations
need to have at least CLASS retention to be visible in the class
files. The benefit is much faster compilation.

### CODE REVIEW

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/AggregatingIncrementalAnnotationProcessingIntegrationTest.groovy [hunk 1]
```diff
@@ -204,4 +208,11 @@ class AggregatingIncrementalAnnotationProcessingIntegrationTest extends Abstract
         expect:
         succeeds "compileJava"
     }
+
+    private boolean serviceRegistryReferences(String... services) {
```
Comment (blindpirate) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/AggregatingIncrementalAnnotationProcessingIntegrationTest.groovy]: I think this method should also test `ServiceRegistry.class` file, not only `ServiceRegistry.java`.

**Comment by blindpirate** [comment]
Oh, I realized this limitation only happens when a annotation processor declared `META-INF/services/javax.annotation.processing.Processor`. That's okay. Forget the previous comment please.

**Comment by oehme** [comment]
It only happens when it registers itself as incremental and the author didn't read the manual :) I'll look if we can fail more quickly in the common case.

**Comment by oehme** [comment]
We now validate the set of annotations passed to the processor.

**Comment by stephanenicolas** [comment]
Thx for this enhancement @oehme . I think it would be nice to add a note about the importance of not using `RUNTIME` retention for annotations on Android. If you do so, all these annotations and more importantly ***all classes using them*** will be kept in the first dex, even in a multidex scenario. And this often breaks dexing.

**Comment by stephanenicolas** [comment]
Also @oehme I am wondering what happens to APs who would like to manage their state by themselves. Is there a way to say Gradle basically: "consider this AP incremental and don't manage anything, I will do it" ?

More generally speaking I guess that the "new M2" releases, simply lets javac expose the elements in files that were previsouly compiled. Javac will expose them to the compiler as elements being annotated. It's nice because it removes any API from incap and simplifies annotation processors. At the same time, it also makes a few optimisations impossible for aggregating APs. We could think of an AP that manages its state itself and would know exactly what elements have changed between two builds. It could fully be incremental and use source retention for its annotation.

This PR looks OK, but I wanted to say that there are some issues in relying on the javac API only and not having one for incap because it will oblige APs to re-process stuff that was already processed, at least partially. It won't be recompiled, ok, but will still be reprocessed, and APs can do better in some cases. For instance, in a recent PR for butterknife, we have found that instead of re-processing a previously compiled parent class, we could simply use the generated class for the parent class if it exists. And the gain in speed is huge as you only have to check one level up, with a very fast check, in a class hierarchy and not to re-process all annotations in all super classes. 

So overall, I couldn't find any real case where this could break but I just wanna say that there are other ways to do things, and APs can legitimately want to do that, especially for optimization purposes.

**Comment by oehme** [comment]
@stephanenicolas I think Android processor authors will know what to do, this information doesn't really belong in a chapter about incremental annotation processing.

We don't intend to offer any kind of state management API. Incremental processing is intended to be as transparent as possible. 

I don't understand the Butterknife example. As far as I understand Butterknife is an isolating processor, so it's already efficient.

**Comment by stephanenicolas** [comment]
About the butterknife example: 
It's not exactly a 1-1 relation, because multiple classes are used to generate one class, but they all belong to the reachable ASTs of one class being processed.

So, when a class using BK extends from other classes using BK, all of them contribute to information to generate the output. A trivial implementation would be, when processing a class, to re-scan all the classes up the hierarchy, but our PR doesn't do that: instead, it reuses the generated classes for the super class. Basically the generated output for a subclass will call the generated output for a super class, and we can check this generated ouput for the super class to guess what annotations were present in the super class. This is more optimal as it limits the re-processing of annotations already processed: it only queries one level up, and not all ancestors, and doesn't even re-process the parent fully to scan for its annotations.

**Comment by oehme** [comment]
> they all belong to the reachable ASTs of one class being processed.

That is the definition of isolating.

The Butterknife PR you refer to doesn't "keep state". It just uses information visible through the AST instead of looking at the RoundEnvironment. That is a perfectly normal thing to do for an isolating processor. The documentation clearly states: You can look at anything you can reach through the element's AST.

> Btw, I wouldn't know how to declare my AP if it would actually want to manage all of its state itself and would like to play nicely with Gradle, not letting gradle manage anything but still not wanting to slow down the build.

I don't see what such a processor would look like. If you have one, please point me to it and tell me why it can't be isolating.

**Comment by stephanenicolas** [comment]
In the PR we did for BK, we look at things that are not reachable through the AST of the class being processed, and it is state information from a previous round: we look at the generated code for an element that is reachable but might have been processed earlier. But what we look at is not directly reachable, and we do this for optimization purposes.

I believe it will become a very common pattern for APs: if you generate code and handle inheritance, you try to generate code for a subclass that will call the super generated class. That's already a pattern for quite a few APs. What will become more of a pattern is that to handle both incremental compilation and modularization, you can "check if there was some code previsouly generated for the super class, and depending on the result, change the output for the subclass.

For the last part of your question: let's imagine an AP that is not isolating and saves its state to a json file: which files have been processed and enough information to know what was generated. This AP could handle it's state of its own. I don't know of such an AP but I think it can really make sense for complex APs like databinding or dagger if they wanted to handle incrementality efficiently.

**Comment by Alexander--** [comment]
@oehme This is a great optimization, but there are some notable differences between processing source code and binary classes. In particular, binary files might not have *argument names*.

**Comment by oehme** [comment]
Thanks @Alexander--! Are you aware of other major differences that people might depend on? In the case of parameter names the processor author could advise their users to use `-parameters`. Or we could add another category that uses the behavior we previously had (recompiling from source).

**Comment by Alexander--** [comment]
@oehme 

> Or we could add another category that uses the behavior we previously had (recompiling from source).

In that case, instead of category, that should probably be a tag/toggle.

> Are you aware of other major differences

That's not directly related (those processors aren't going to work under current model anyway), but there is also a matter of direct AST access, such as via `com.sun.source.util.Trees`. Trying to use `Trees` should cleanly fail with IllegalArgumentException (because you are replacing `ProcessingEnvironment` impleementation, used by `Trees#instance`), but attempting to bluntly cast interfaces to their javac-internal implementations may still provide a way to screw around with AST.

You should explicitly mention in documentation, that processors like ButterKnife, Lombok and Checker Framework, that rely on internal APIs, might not be compatible with incremental annotation processing.

**Comment by oehme** [comment]
Thanks @Alexander--, I'll revamp the documentation before 4.7 goes out.

**Pull request merged**